### ISSUE TITLE
Change verbosity flag `-v 1` -> `-v`

### ISSUE THREAD

**Issue by kousu** [neuropoly/spinalcordtoolbox#2676]
It's more conventional in unix to pass `--verbose` or `-v` as a flag, not an option, i.e. it's existence determines whether it applies, not an argument passed after it.

Our wiki [used to suggest](https://github.com/neuropoly/spinalcordtoolbox/wiki/variables/_compare/97006a885dde276c9e8469f3f69b0089639eb609...0de1bc657993c94740da8db96a21cf162b8a49c5) an argument of `-v 0` through `-v 2` to control log level. I've changed this in `sct_testing` to be just `` for "off" or `-v` for "on". We should go through the rest of our CLIs and make them conform.

We can use [`action="count"`](https://docs.python.org/3/library/argparse.html#action) so you can do `$script -v -v -v` or `$script --verbose -vv` or `$script -vvv`

Ref: https://github.com/neuropoly/spinalcordtoolbox/pull/2672#issuecomment-619124796_

**Comment by joshuacwnewton** [comment]
Hello! I'm interested in taking on this issue.

> We should go through the rest of our CLIs and make them conform.

Just to confirm, it looks like each script in `scripts/` implements its own `get_parser()`, each with separate verbosity options. And, you'd like each one to be updated to follow this new behaviour, right?

A more long-term solution could be to use a centralized argparser that contains arguments common to each parser. Then, have each script add its own args to the base parser. Something similar to [this](https://stackoverflow.com/questions/15494574/using-pythons-argparse-in-multiple-scripts-backed-by-multiple-custom-modules), possibly? That way there's less boilerplate, and less need to tweak the same bits of code in many different places. That would be a bigger change, though, and affect much more than just verbosity, so it might be better for a separate issue.

Either way, I'm happy to go with what you feel is best. Let me know what you think, and I can get started on a PR. :)

**Comment by jcohenadad** [comment]
@joshuacwnewton thank you so much for offering to help!  🎉

As you rightfully said, some CLIs are still using our custom parser via `msct_parser`, and we are (slowly) transitioning to Python's `argparse` https://github.com/neuropoly/spinalcordtoolbox/issues/1548. I think it would make more sense to first finalize the transition to a 100% argparse, and then change verbose behaviour. Would that be OK with you to work on that?

**Comment by joshuacwnewton** [comment]
Of course! Thank you @jcohenadad for clearing things up, as well as linking the relevant issue. I'll spend some time getting familiar with the previous work done on the argparse conversion (e.g. #2315, #2337) then we can continue chatting on the main issue, if that sounds good. :slightly_smiling_face:

**Comment by joshuacwnewton** [comment]
In addition to what @kousu says, something to keep in mind is that the verbosity isn't consistent between scripts, either.

Some have 2 verbose levels ('0' and '1') while others have 3 verbose levels ('0', '1', and '2'). Argparse description strings are also inconsistent: sometimes it's called `'off'`/`'on'`, sometimes it's called `'none'`/`'classic'`/`'extended'`, and sometimes the actual functionality is referenced, see below.

https://github.com/neuropoly/spinalcordtoolbox/blob/fce82d8c98fab1cc878c4e8c48fa1c515f0db35c/scripts/sct_flatten_sagittal.py#L123-L128

**Comment by jcohenadad** [comment]
> Might be good to hammer down how we want to communicate verbosity, and what types of output correspond to each level.

💯  for the sake of simplicity, i would go with a single -v argument, that would be `store_true`, and the would bring logging level from INFO to DEBUG.

**Comment by joshuacwnewton** [comment]
> for the sake of simplicity, i would go with a single -v argument, that would be store_true, and the would bring logging level from INFO to DEBUG.

@jcohenadad This makes sense to me, given that these are user-facing scripts! :)

I've also updated the Wiki to clarify this information a little by adding a section called **How logging appears to users of SCT scripts**: https://github.com/neuropoly/spinalcordtoolbox/wiki/Logging

**Comment by joshuacwnewton** [comment]
While looking at verbosity across SCT, I noticed that there are many places that rely on `verbose == 2`. Example:

https://github.com/neuropoly/spinalcordtoolbox/blob/bf3de687c20ac86717de3584296af1931e247784/scripts/sct_utils.py#L295-L298

I can simplify these cases to be simply `if verbose:` if our goal is to have verbosity be consistent across SCT. 

But, if the distinction between `verbose==1` and `verbose==2` _does_ matter, then `action="store_true"` may not be sufficient.

**Comment by jcohenadad** [comment]
for these cases i would merge verbose 1 and 2, ie: if -v, then do what is currently under verbose 1 and 2

**Comment by joshuacwnewton** [comment]
I've noticed that some scripts have verbose output turned on by default. e.g. 

https://github.com/neuropoly/spinalcordtoolbox/blob/bf3de687c20ac86717de3584296af1931e247784/scripts/sct_dmri_separate_b0_and_dwi.py#L31-L37

With `action="store_true"`, verbose would be `False` by default. So, `-v` would have to be manually specified to have the full verbose output (rather than it being on by default.)

Whether this matters seems like it would be case-by-case depending on the script, but I thought I would mention it here for posterity.

**Comment by joshuacwnewton** [comment]
I'm realizing now that this proposal has very similar issues as #2843. Many of our scripts have verbosity turned on by default, so `action=store_true` doesn't quite work nicely. Using 0/1 did indeed have a purpose: it allowed us to specify `-v 0` to turn off verbose output.

I believe the suggestion made by @jcohenadad in https://github.com/neuropoly/spinalcordtoolbox/issues/2843#issuecomment-694586745 to harmonize usage with `0/1 (int)` rather than `"0"/"1"/"2" (str)` is appropriate here, too. And, I feel that switching to using `action="store_true"` doesn't make as much sense anymore.

**Comment by jcohenadad** [comment]
Indeed, but in this case there is an additional layer of complexity, because we might not _need_ the verbose=0 as initially designed. When I first designed this v=0, 1, 2 i had in mind the scenario where users would like no terminal output at all (v=0), but I realize now that this option seems irrelevant (ie: what's the problem with having those outputs?). So, we could also change the current behavior, and have: 
- v=0: logging.level=info (default behaviour)
- v=1: logging.level=debug, with output of plotting and images for debugging purposes

or, we stay closer with the current design, and go for:
- v=0: logging.level=warning, ie minimum outputs.
- v=1:  logging.level=info (default behaviour)
- v=2: logging.level=debug, with output of plotting and images for debugging purposes

**Comment by joshuacwnewton** [comment]
> So, we could also change the current behavior, and have:
> * v=0: logging.level=info (default behaviour)
> * v=1: logging.level=debug, with output of plotting and images for debugging purposes

This is basically the same as what you proposed in https://github.com/neuropoly/spinalcordtoolbox/issues/2676#issuecomment-671135177, right? (INFO -> DEBUG)

This is what I had been following in my recent commits, so these changes are already done, oops. :sweat_smile:

**Comment by jcohenadad** [comment]
> > So, we could also change the current behavior, and have:
> > 
> > * v=0: logging.level=info (default behaviour)
> > * v=1: logging.level=debug, with output of plotting and images for debugging purposes
> 
> This is basically the same as what you proposed in [#2676 (comment)](https://github.com/neuropoly/spinalcordtoolbox/issues/2676#issuecomment-671135177), right? (INFO -> DEBUG)
> 
> This is what I had been following in my recent commits, so these changes are already done, oops. 😅

Yes, you're absolutely right, this is what I have been proposing in [#2676 (comment)](https://github.com/neuropoly/spinalcordtoolbox/issues/2676#issuecomment-671135177). I'm sorry to bring more confusion to this topic, but i'm still unsure as what is the best strategy here 😅

**Comment by joshuacwnewton** [comment]
> I'm sorry to bring more confusion to this topic, but i'm still unsure as what is the best strategy here :sweat_smile: 

No worries! This is partially my fault for going ahead and making changes before we had settled on a choice. :stuck_out_tongue_closed_eyes:

**Comment by joshuacwnewton** [comment]
That was completely accidental, my apologies!

**Comment by joshuacwnewton** [comment]
> I'm sorry to bring more confusion to this topic, but i'm still unsure as what is the best strategy here sweat_smile

My apologies, as well. I got a bit confused myself with my recent changes, and I see what you mean now about 0/1/2 being an appealing choice, as well.

**Comment by joshuacwnewton** [comment]
Very sorry. There was a misunderstanding on my part. 

* My comment, https://github.com/neuropoly/spinalcordtoolbox/issues/2676#issuecomment-695076627, was incorrect. Verbose is not on by default. I misunderstood.
* Julien's original suggestion, https://github.com/neuropoly/spinalcordtoolbox/issues/2676#issuecomment-671135177 is still good. `action='store_true'` is still a viable solution.

My apologies for wasting time. I'll demonstrate what I mean in a new PR.

**Comment by joshuacwnewton** [comment]
SCT's current usage of verbosity is more complicated than I realized. It's used in subtly different ways in different parts of the project, e.g. `sct.printv` and parts of the core API, e.g. `centerline/core.py`, `deepseg_legion/core.py`, etc.)

I think trying to standardize this parameter requires a much deeper investment of time than I first thought. I'm not sure this is high-enough priority to warrant the effort. I can make a draft PR with what I have, but I'm not sure if it's worth finishing considering there are other high-priority tasks.

**Comment by joshuacwnewton** [comment]
Removing this from the 5.0.0 milestone as to not hold up the release. If the remaining 5.0.0 tasks finish quickly and there's time to spare, perhaps it could be added back, but we can discuss that more in the upcoming Tuesday meeting. :slightly_smiling_face:

**Comment by joshuacwnewton** [comment]
Removing `good first issue` as to not mislead any new contributors (e.g. Hacktoberfest). This has become more complicated (and is lower priority compared to some other tasks).

**Comment by Drulex** [comment]
I know I'm late to the party, but I personally always use
```
usage: [...]
--log-level LOG_LEVEL     Logging level (eg. INFO, DEBUG, WARNING. See Python logging docs). Default INFO.
```

in my python CLI applications. It's simple and intuitive. There is no guessing of what the levels mean or what the default level is.

**Comment by jcohenadad** [comment]
i like that approach https://github.com/neuropoly/spinalcordtoolbox/issues/2676#issuecomment-723089582, it does indeed have the merit of being intuitive. The downside is that it implies large breakage in cross-compatibility.

**Comment by Drulex** [comment]
> The downside is that it implies large breakage in cross-compatibility.

You can always support both for backwards compatibility and eventually deprecate one. Maybe print a warning when users use `-v` and say it will be deprecated in v6.0 or something. Meanwhile all docs use `--log-level`. It is a widely used approach.

**Comment by kousu** [comment]
#3139 makes it easier to handle these generic parsers by centralizing a bunch of things in a `SCTArgumentParser` class.

I have a suggestion for how to handle moving all the redundant `--help` definitions into it at https://github.com/neuropoly/spinalcordtoolbox/pull/3139/files#r549895535:

```

    def add_argument_group(self, title, *args, **kwargs):
         g = [g for g in parser._action_groups if g.title == title]
         if g:
             return g
         else:
             return super().add_argument_group(title, *args, **kwargs)         
```

that should work to let you centralize `-h`, `-v`, `-r`, or anything else. And we can subclass SCTArgumentParser if there are subsets of scripts that need further shared parts that the rest of the scripts don't share.

**Comment by SomeoneInParticular** [comment]
I actually was going to do this as part of a general rewrite as part of a broader rewrite of the argument parser, in preparation to solve #4801. Will take a while though (lots of technical debt, with verbosity checks strewn across the entire codebase), but I'll link a PR which will resolve this once I push the first commit.

**Comment by SomeoneInParticular** [comment]
So, before I start implementing this in any capacity, I wanted to run the proposed changes to ensure their aren't any major objections to anything:

**New `logging` Utility**

Primarily used to isolate logging code and logic from where it currently is (in `sys`), as multiple functions will need to be added to consolidate things in one place:

* Return the current logging (verbosity)
* Set logging to be quiet (`logging` level == 30, low verbosity) 
* Set logging to be verbose (`logging` level == 10, high verbosity)
* Print messages of various types (debug, info, warning, error etc.)
* Define where to put the logs (i.e. to file, or to console)

I'm leaning towards a `Logger.logging` like setup, where the user just needs to import something like `sct_logging` and leave the rest to the manager, but I am open to further discussion on this matter.

**Changes to the -v argument:**

* `-v 0` -> `-q`; if `-q` is present (“quite”), logging it set to WARNING mode (only warnings and messages). 
    * I have confirmed no CLI script uses `-q` yet; yippee!
* `-v 1` -> Nothing; if neither the above or below arguments are present, logging is set to INFO mode (prior, plus updates on program status/progress)
* `-v 2` -> `-v`; if `-v` is present (“verbose”), logging is set to DEBUG mode (very detailed updates on program status and progress)

Both `-v` and `-q` would just call the corresponding actions in the `logging` utility, making their functionality universal to all CLI commands. Likewise, post-CLI scripts can reference the verbosity through the utility as well.

**Comment by joshuacwnewton** [comment]
> I'm leaning towards a `Logger.logging` like setup, where the user just needs to import something like `sct_logging` and leave the rest to the manager, but I am open to further discussion on this matter.

Does this imply a global logger + logging state? If so, then one caveat to keep in mind is the leaking of state when one CLI script calls another. Currently, we do actually use `Logger.logging` in the Python API, and so we ran into that issue once prior: 

- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3341

> Primarily used to isolate logging code and logic from where it currently is (in sys), as multiple functions will need to be added to consolidate things in one place:

If we're consolidating state in one place, does that mean that the plan is to replace the existing verbosity settings throughout SCT?

- **`Logging.logger`**: Calls to `logger.info`, `logger.warning`, `logger.error`, etc. in the Python API that rely on `Logging`'s loglevel system.
- **`verbose`**: Parameters that get manually passed from function to function, including:
  - Conditional statements `if verbose`, `if verbose == 2`, etc. that control print statements
  - Printing functions that take `verbose` as an argument (e.g. `printv(verbose=)`)

**Comment by SomeoneInParticular** [comment]
> Does this imply a global logger + logging state? If so, then one caveat to keep in mind is the leaking of state when one CLI script calls another. Currently, we do actually use Logger.logging in the Python API, and so we ran into that issue once prior:

Yes, there would be a global verbosity state, which would be used as the "main" verbosity . That is, for logging which is currently done using (`printv`)[https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/master/spinalcordtoolbox/utils/sys.py#L352], as well as the "default" verbosity if no local verbosity was defined for this scope (described as follows). 

There will also be "local" verbosity, initialized on request managed by a (global) dictionary, to prevent verbosity leakage in the handful of cases where we want distinct verbosity levels. 

I'm also going to explore using (`inspect.stack`)[https://docs.python.org/3/library/inspect.html#inspect.stack] to manage this (as it would prevent use from needing to always pass the current module's `__name__` every time), but that needs to be tested before I can confirm it works fully.


> If we're consolidating state in one place, does that mean that the plan is to replace the existing verbosity settings throughout SCT?

Yes, specifically by moving `verbose` checks to be managed implicitly by `logger`-like function calls. 

For example, this call:

`printv(f'\nSlice {slice_orig},  tSNR = {tsnr_roi:.2f}', type='info')`

would become this:

`sct_logging.info(f'\nSlice {slice_orig},  tSNR = {tsnr_roi:.2f}')`

And 

`if self.param.verbose == 2:`

would become:

`if sct_logging.is_verbose():`

**Comment by joshuacwnewton** [comment]
Thanks for the clarification!! I think this would be a welcome change... albeit a very tedious and time-consuming overhaul given how many places that `verbose` and `Logging.logger.{debug/info/warning/error}` appear in the SCT code base. (My past wrangling with this issue as a fresh-faced dev still haunts me, see https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2676#issuecomment-695141546 and https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2897#issue-704728926. 😅)

One more Q:

- What part of this issue specifically is blocking [#4801](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4801)?
- Or, in other words, would it be difficult to implement [#4801](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4801) using our current logging setup, then transition the time logging to the proposed "`-q` + `-v` + backend overhaul" after the fact?

Just curious about your vision for #4801 and how it intersects with this proposal. 😊

**Comment by SomeoneInParticular** [comment]
I don't mind a tedium task to whittle away at between other complex projects, honestly; just means it will take a while to shift the corresponding PR out of draft form.

> What part of this issue specifically is blocking https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4801?

Nothing; sorry for the confusion. What was blocking #4801 was that we didn't have arguments set up yet; this should be resolved with #4819.

In fact, I plan on using the PR which will resolve #4801 as "demo" of how universal CLI functionality (such as the `-v` overhaul discussed here) could be implemented into SCT.
